R/output.R

```
@@ -288,15 +288,19 @@ process_file = function(text, output) {
   # when in R CMD check, turn off the progress bar (R-exts said the progress bar
   # was not appropriate for non-interactive mode, and I don't want to argue)
   progress = opts_knit$get('progress') && !is_R_CMD_check()
+  labels = unlist(lapply(groups, function(g) {
+    if (is.list(g$params)) g[[c('params', 'label')]] else ''
+  }))
   if (progress) {
-    labels = unlist(lapply(groups, function(g) {
-      if (is.list(g$params)) g[[c('params', 'label')]] else ''
-    }))
     pb_fun = getOption('knitr.progress.fun', txt_pb)
     pb = if (is.function(pb_fun)) pb_fun(n, labels)
     on.exit(if (!is.null(pb)) pb$done(), add = TRUE)
   }
   wd = getwd()
+  if (xfun::pkg_available("rlang", "1.0.0")) {
+    rlang::global_entrace(TRUE)
```

I don't think we can do better currently, but this will turn it off for folks who previously had it turned on before running this this. I filed an issue in rlang: https://github.com/r-lib/rlang/issues/1552.

---

Note that this will fail if the function is called within a `tryCatch()` or `withCallingHandlers()`. That's because `base::globalCallingHandlers()` requires an empty handler stack. For this reason `withCallingHandlers(error = rlang::cnd_entrace)` would be better if possible.

---

This should at least be guarded behind an error-swallowing `tryCatch()`. This way when `knit()` is called in testthat unit tests for instance, the error thrown by `globalCallingHandlers()` will not interrupt the knitting process.

---

ohhh If guarded in a `tryCatch()`, it will always fail!

So I think there is no reliable way to call `globalCallingHandlers()` except in an RProfile or directly in the console...

---

I've just realised that with the work to make `global_entrace()` register knitr handlers in https://github.com/r-lib/rlang/pull/1557, all that is needed is to call it at an appropriately late time in the knitr initialisation process: When `knitr.in.progress` is `TRUE` and `knitr::opts_chunk$set(calling.handlers = )` is available. It works because all errors get entraced, including unexpected ones.

---

Though doing that would mean delegating the sewing of _all_ errors and warnings to rlang by default, which doesn't seem appropriate and probably has unintended consequences (e.g. in terms of performance). So the `withCallingHandlers(error = function(cnd) rlang::entrace(cnd))` approach is best.

---

@cderv is it obvious to you what the next step is?

---

Yes I think I understood. I made some changes according to that for a nested `withCallingHandlers()`